{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Are my twin brother and I actually identical twins? \n",
    "   \n",
    "Ever since we were born my twin brother Paul and I have been told that we were\n",
    "fraternal twins. However, recently my mother has taken up genealogy as a hobby \n",
    "and had the entire family do a DNA test using the FamilyTreeDNA (FTDNA). The DNA testing services presents the results of the tests in terms of shared [**centimorgans**](https://en.wikipedia.org/wiki/Centimorgan) with other users of their service.  For geneaological purposes this is sufficient as it allows the service to make an imperfect guess regarding the relationship of two people [(they often use other factors such as self-reported age in these estimates)](https://slate.com/technology/2019/10/23andme-family-secrets-half-siblings-cousins.html). [ISSOG](https://isogg.org/wiki/Autosomal_DNA_statistics) provides a nifty chart of the relationship between centimorgans and relationship:    \n",
    "\n",
    "<img src=\"img/Shared_cM.jpg\">\n",
    "\n",
    "The problem is that the highest match, for where people share approximately 50% of their DNA, is parent/child.  This is how my brother and I are shown to be related.  What I want to confirm is that we are indeed fully identical.  Fortunately, FamilyTreeDNA does make the raw genetic data available for download and I asked my parents and brother to send me copies of their data.  If Paul and I are identical we should share almost 100% of our genetic code.  Furthermore, we should share significantly more with each other than with either of our parents with whom we should share on average about 50%.     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importing the libraries we'll use in this notebook.\n",
    "import matplotlib.pylab as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from functools import reduce"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading in the Data \n",
    "\n",
    "\n",
    "Ok, huge disclaimer - I'm not a geneticist or a biologist but this is what I think things are. \n",
    "\n",
    "The data files from FTDNA inludes:  \n",
    "\n",
    "+ **RSID (Reference SNP ID)** - An SNP is a [single-nucleotide polymorphism](https://en.wikipedia.org/wiki/Single-nucleotide_polymorphism) which, as I understand it, are the nucleotides that vary from person to person and are not the same for every human being.\n",
    "\n",
    "+ **Chromosome** - The chromosome where a given SNP is found. \n",
    "\n",
    "+ **Position** - The position on the chromosome where an SNP is located. \n",
    "\n",
    "+ **Result** - The paired allele values for the SNP. Not present when allele1/allele2 columns are provided. \n",
    "\n",
    "+ **allele1/allele2** - The individual values for alleles for a given SNP.  Not present when a result column is provided.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/scott/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n",
      "/home/scott/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scott\n",
      "        # name chromosome  position allele1 allele2\n",
      "0   rs1042522         17   7579472       G       G\n",
      "1   rs1057909         10  96741051       A       A\n",
      "2   rs1175544          3  12467044       T       C\n",
      "3  rs12721628          7  99367939       G       G\n",
      "4  rs12721636          7  99381766       C       C \n",
      "\n",
      "Paul\n",
      "        # name chromosome  position allele1 allele2\n",
      "0   rs1042522         17   7579472       G       G\n",
      "1   rs1057909         10  96741051       A       A\n",
      "2   rs1175544          3  12467044       T       C\n",
      "3  rs12721628          7  99367939       G       G\n",
      "4  rs12721636          7  99381766       C       C \n",
      "\n",
      "Mom\n",
      "          RSID CHROMOSOME  POSITION RESULT\n",
      "0   rs4477212          1     82154     AA\n",
      "1   rs3094315          1    752566     --\n",
      "2   rs3131972          1    752721     GG\n",
      "3  rs12562034          1    768448     --\n",
      "4  rs12124819          1    776546     -- \n",
      "\n",
      "Dad\n",
      "          RSID CHROMOSOME POSITION RESULT\n",
      "0   rs4477212          1    82154     AA\n",
      "1   rs3094315          1   752566     --\n",
      "2   rs3131972          1   752721     AG\n",
      "3  rs12562034          1   768448     --\n",
      "4  rs12124819          1   776546     --\n"
     ]
    }
   ],
   "source": [
    "# Reading in the csv files for each person.\n",
    "scott = pd.read_csv('scott_dna.csv')\n",
    "paul = pd.read_csv('paul_dna.csv')\n",
    "\n",
    "mom = pd.read_csv('mom_dna.csv')\n",
    "dad = pd.read_csv('dad_dna.csv')\n",
    "\n",
    "# Peeking at our data. \n",
    "\n",
    "print(\"Scott\\n\",\n",
    "      scott.head(), \n",
    "      '\\n\\nPaul\\n', \n",
    "      paul.head(), \n",
    "      '\\n\\nMom\\n', \n",
    "      mom.head(),\n",
    "      '\\n\\nDad\\n',\n",
    "      dad.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cleaning the Data \n",
    "\n",
    "As you can, the data files are formatted slightly differently so that Paul and I lack a \"Result\" column similar to that found in my parents data. However, we can construct the result column from the \"allele1\" and \"allele2\" column.  \n",
    "\n",
    "Another issue is testing error.  For example, some SNPs are listed on the non-existant chromosome 0. Likewise, when an explicit error has been made the result is listed as \"--\".  In both cases we will drop these SNPs from the analysis. \n",
    "\n",
    "Finally, we'll relabel our columns so that we can tell individuals apart and have consistent naming conventions. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's create a combined result column like my parents' data.\n",
    "scott['result'] = scott['allele1'] + scott['allele2']\n",
    "paul['result'] = paul['allele1'] + paul['allele2']\n",
    "\n",
    "# Drop out chromosomes that equal zero - this is a testing error.\n",
    "scott.drop(scott[scott.chromosome == 0].index, inplace=True)\n",
    "paul.drop(paul[paul.chromosome == 0].index, inplace=True)\n",
    "mom.drop(mom[mom.CHROMOSOME == 0].index, inplace=True)\n",
    "dad.drop(dad[dad.CHROMOSOME == 0].index, inplace=True)\n",
    "\n",
    "# Drop cases where there is no result\n",
    "mom.drop(mom[mom.RESULT == '--'].index, inplace=True)\n",
    "scott.drop(scott[scott.result == '--'].index, inplace=True)\n",
    "paul.drop(paul[paul.result == '--'].index, inplace=True)\n",
    "dad.drop(dad[dad.RESULT == '--'].index, inplace=True)\n",
    "\n",
    "# Drop out \"famfinder\" - combined file header\n",
    "scott.drop(scott[scott['# name'] == '# famfinder'].index, inplace=True)\n",
    "paul.drop(paul[paul['# name'] == '# famfinder'].index, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "scott.columns = [\"RSID\",\n",
    "                 \"s_chromosome\",\n",
    "                 \"s_position\",\n",
    "                 \"s_allele1\",\n",
    "                 \"s_allele2\",\n",
    "                 \"s_result\"]\n",
    "\n",
    "paul.columns = [\"RSID\",\n",
    "                \"p_chromosome\",\n",
    "                \"p_position\",\n",
    "                \"p_allele1\",\n",
    "                \"p_allele2\",\n",
    "                \"p_result\"]\n",
    "\n",
    "mom.columns = [\"RSID\",\n",
    "               \"m_chromosome\",\n",
    "               \"m_position\",\n",
    "               \"m_result\"]\n",
    "\n",
    "dad.columns = [\"RSID\",\n",
    "               \"d_chromosome\",\n",
    "               \"d_position\",\n",
    "               \"d_result\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Joining Data \n",
    "\n",
    "Now we need to join the data so it can be uniformly compared.  We'll use my mothers data as a base and join in the basis of the RSID.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = [mom, dad, scott, paul]\n",
    "df = reduce(lambda left, right: pd.merge(left, right, on='RSID'), dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Since dad won't share an X chromosome with us, let's drop it to put \n",
    "# both my parents on an equal footing. Only autosomal \n",
    "\n",
    "df.drop(df[df['m_chromosome'] == 'X'].index, inplace=True)\n",
    "df.drop(df[df['d_chromosome'] == 'X'].index, inplace=True)\n",
    "df.drop(df[df['p_chromosome'] == 'X'].index, inplace=True)\n",
    "df.drop(df[df['s_chromosome'] == 'X'].index, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RSID</th>\n",
       "      <th>m_chromosome</th>\n",
       "      <th>m_position</th>\n",
       "      <th>m_result</th>\n",
       "      <th>d_chromosome</th>\n",
       "      <th>d_position</th>\n",
       "      <th>d_result</th>\n",
       "      <th>s_chromosome</th>\n",
       "      <th>s_position</th>\n",
       "      <th>s_allele1</th>\n",
       "      <th>s_allele2</th>\n",
       "      <th>s_result</th>\n",
       "      <th>p_chromosome</th>\n",
       "      <th>p_position</th>\n",
       "      <th>p_allele1</th>\n",
       "      <th>p_allele2</th>\n",
       "      <th>p_result</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs4477212</td>\n",
       "      <td>1</td>\n",
       "      <td>82154</td>\n",
       "      <td>AA</td>\n",
       "      <td>1</td>\n",
       "      <td>82154</td>\n",
       "      <td>AA</td>\n",
       "      <td>1</td>\n",
       "      <td>82154</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>AA</td>\n",
       "      <td>1</td>\n",
       "      <td>82154</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>AA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs3131972</td>\n",
       "      <td>1</td>\n",
       "      <td>752721</td>\n",
       "      <td>GG</td>\n",
       "      <td>1</td>\n",
       "      <td>752721</td>\n",
       "      <td>AG</td>\n",
       "      <td>1</td>\n",
       "      <td>752721</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>AG</td>\n",
       "      <td>1</td>\n",
       "      <td>752721</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs11240777</td>\n",
       "      <td>1</td>\n",
       "      <td>798959</td>\n",
       "      <td>GG</td>\n",
       "      <td>1</td>\n",
       "      <td>798959</td>\n",
       "      <td>AG</td>\n",
       "      <td>1</td>\n",
       "      <td>798959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>AG</td>\n",
       "      <td>1</td>\n",
       "      <td>798959</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs4970383</td>\n",
       "      <td>1</td>\n",
       "      <td>838555</td>\n",
       "      <td>CC</td>\n",
       "      <td>1</td>\n",
       "      <td>838555</td>\n",
       "      <td>AC</td>\n",
       "      <td>1</td>\n",
       "      <td>838555</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>CC</td>\n",
       "      <td>1</td>\n",
       "      <td>838555</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs4475691</td>\n",
       "      <td>1</td>\n",
       "      <td>846808</td>\n",
       "      <td>CC</td>\n",
       "      <td>1</td>\n",
       "      <td>846808</td>\n",
       "      <td>CC</td>\n",
       "      <td>1</td>\n",
       "      <td>846808</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>CC</td>\n",
       "      <td>1</td>\n",
       "      <td>846808</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>CC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs7537756</td>\n",
       "      <td>1</td>\n",
       "      <td>854250</td>\n",
       "      <td>AG</td>\n",
       "      <td>1</td>\n",
       "      <td>854250</td>\n",
       "      <td>AA</td>\n",
       "      <td>1</td>\n",
       "      <td>854250</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>AG</td>\n",
       "      <td>1</td>\n",
       "      <td>854250</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs13302982</td>\n",
       "      <td>1</td>\n",
       "      <td>861808</td>\n",
       "      <td>GG</td>\n",
       "      <td>1</td>\n",
       "      <td>861808</td>\n",
       "      <td>GG</td>\n",
       "      <td>1</td>\n",
       "      <td>861808</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>GG</td>\n",
       "      <td>1</td>\n",
       "      <td>861808</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>GG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rs1110052</td>\n",
       "      <td>1</td>\n",
       "      <td>873558</td>\n",
       "      <td>GT</td>\n",
       "      <td>1</td>\n",
       "      <td>873558</td>\n",
       "      <td>GT</td>\n",
       "      <td>1</td>\n",
       "      <td>873558</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>TG</td>\n",
       "      <td>1</td>\n",
       "      <td>873558</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>TG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs2272756</td>\n",
       "      <td>1</td>\n",
       "      <td>882033</td>\n",
       "      <td>AG</td>\n",
       "      <td>1</td>\n",
       "      <td>882033</td>\n",
       "      <td>AG</td>\n",
       "      <td>1</td>\n",
       "      <td>882033</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>AG</td>\n",
       "      <td>1</td>\n",
       "      <td>882033</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>AG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rs17160698</td>\n",
       "      <td>1</td>\n",
       "      <td>887162</td>\n",
       "      <td>TT</td>\n",
       "      <td>1</td>\n",
       "      <td>887162</td>\n",
       "      <td>TT</td>\n",
       "      <td>1</td>\n",
       "      <td>887162</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>TT</td>\n",
       "      <td>1</td>\n",
       "      <td>887162</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>TT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         RSID m_chromosome  m_position m_result d_chromosome d_position  \\\n",
       "0   rs4477212            1       82154       AA            1      82154   \n",
       "1   rs3131972            1      752721       GG            1     752721   \n",
       "2  rs11240777            1      798959       GG            1     798959   \n",
       "3   rs4970383            1      838555       CC            1     838555   \n",
       "4   rs4475691            1      846808       CC            1     846808   \n",
       "5   rs7537756            1      854250       AG            1     854250   \n",
       "6  rs13302982            1      861808       GG            1     861808   \n",
       "7   rs1110052            1      873558       GT            1     873558   \n",
       "8   rs2272756            1      882033       AG            1     882033   \n",
       "9  rs17160698            1      887162       TT            1     887162   \n",
       "\n",
       "  d_result s_chromosome s_position s_allele1 s_allele2 s_result p_chromosome  \\\n",
       "0       AA            1      82154         A         A       AA            1   \n",
       "1       AG            1     752721         A         G       AG            1   \n",
       "2       AG            1     798959         A         G       AG            1   \n",
       "3       AC            1     838555         C         C       CC            1   \n",
       "4       CC            1     846808         C         C       CC            1   \n",
       "5       AA            1     854250         A         G       AG            1   \n",
       "6       GG            1     861808         G         G       GG            1   \n",
       "7       GT            1     873558         T         G       TG            1   \n",
       "8       AG            1     882033         A         G       AG            1   \n",
       "9       TT            1     887162         T         T       TT            1   \n",
       "\n",
       "  p_position p_allele1 p_allele2 p_result  \n",
       "0      82154         A         A       AA  \n",
       "1     752721         A         G       AG  \n",
       "2     798959         A         G       AG  \n",
       "3     838555         C         C       CC  \n",
       "4     846808         C         C       CC  \n",
       "5     854250         A         G       AG  \n",
       "6     861808         G         G       GG  \n",
       "7     873558         T         G       TG  \n",
       "8     882033         A         G       AG  \n",
       "9     887162         T         T       TT  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Visually inspect the data - I want a few more rows than df.head() \n",
    "df[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculation Percent Agreement \n",
    "\n",
    "Let's calculate raw percent agreement.  It should be noted this is **not how geneticists and genetic geneaologists calculate shared DNA** - it's going to over inflated shared DNA between unrelated individuals (**for instance, my mom and dad are not related!***).  But it will let us determine if my brother and I are identical.  \n",
    "\n",
    "I am sure that there is a way to calculate chance adjusted agreement but this will answer our initial question.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate agreement\n",
    "df['s_p_agree'] = df['s_result'] == df['p_result']\n",
    "df['s_m_agree'] = df['s_result'] == df['m_result']\n",
    "df['s_d_agree'] = df['s_result'] == df['d_result']\n",
    "df['p_m_agree'] = df['p_result'] == df['m_result']\n",
    "df['p_d_agree'] = df['p_result'] == df['d_result']\n",
    "df['m_d_agree'] = df['d_result'] == df['m_result']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create an array of comparisons \n",
    "pct_agree = np.array([\n",
    "    [1, df['s_p_agree'].mean(), df['s_m_agree'].mean(), \n",
    "     df['s_d_agree'].mean()],\n",
    "    [df['s_p_agree'].mean(), 1, df['p_m_agree'].mean(),\n",
    "     df['p_d_agree'].mean()],\n",
    "    [df['s_m_agree'].mean(), df['p_m_agree'].mean(),\n",
    "     1, df['m_d_agree'].mean()],\n",
    "    [df['s_d_agree'].mean(), df['p_d_agree'].mean(),\n",
    "     df['m_d_agree'].mean(), 1]\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Visualized the percent agreement data. \n",
    "\n",
    "names = ['Scott', 'Paul', \"Mom\", \"Dad\"]\n",
    "fig, ax = plt.subplots()\n",
    "im = ax.imshow(pct_agree, cmap=plt.cm.ocean)\n",
    "\n",
    "ax.set_xticks(np.arange(len(names)))\n",
    "ax.set_yticks(np.arange(len(names)))\n",
    "\n",
    "ax.set_xticklabels(names)\n",
    "ax.set_yticklabels(names)\n",
    "\n",
    "plt.setp(ax.get_xticklabels(), rotation=45, ha=\"right\",\n",
    "         rotation_mode=\"anchor\")\n",
    "\n",
    "for i in range(len(names)):\n",
    "    for j in range(len(names)):\n",
    "        text = ax.text(j, i, round(pct_agree[i, j], 2),\n",
    "                       ha=\"center\", va=\"center\", color=\"black\")\n",
    "\n",
    "fig.tight_layout()\n",
    "plt.title('Percent Agreement')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above plot is pretty telling!  Paul and I are showing 100% agreement here on our base pairs.  That looks pretty identical.  However, one thing many people don't know is that identical twins are not truly identical - there are subtle mutations that mean that they are almost, but not quite, perfect copies.  Below I report the the number of base pairs on which we differed as well as how many by which I differed from my mother.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "agreement = df.s_p_agree.value_counts().to_frame()\n",
    "agreement[\"s_m_agree\"] = df.s_m_agree.value_counts()\n",
    "agreement"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hamming Distance \n",
    "\n",
    "Another way to calculate difference is a hamming distance.  Hamming distance represents the number of positions that the strings differ from one another.  In this case, rather than looking at agreement between base pairs we're looking at agreement between alleles.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create strings of DNA for comparison. \n",
    "s = ''.join(list(df['s_result']))\n",
    "p = ''.join(list(df['p_result']))\n",
    "m = ''.join(list(df['m_result']))\n",
    "d = ''.join(list(df['d_result']))\n",
    "\n",
    "#let's check to make sure we're comparing apples to apples. \n",
    "print(\"DNA same length?:\", len(s) == len(p) == len(m) == len(d), \"\\n\") \n",
    "\n",
    "def hamming_distance(s1, s2):\n",
    "    return sum(ch1 != ch2 for ch1,ch2 in zip(s1,s2))\n",
    "\n",
    "# Calculate hamming distances. \n",
    "print(\"*** Hamming Distances between Individuals DNA ***\")\n",
    "print(\"Mom w/ Dad:    \", hamming_distance(m, d))\n",
    "print(\"Mom w/ Scott:  \", hamming_distance(m, s))\n",
    "print(\"Dad w/ Scott:  \", hamming_distance(d, s))\n",
    "print(\"Mom w/ Paul:   \", hamming_distance(m, p))\n",
    "print(\"Dad w/ Paul:   \", hamming_distance(d, p))\n",
    "print(\"Scott w/ Paul: \", hamming_distance(s, p))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Conclusion\n",
    "\n",
    "So are Paul and I identical? Identical twins are *not truly identical* and there can be minute differences in DNA resulting from mutation or testing error.  The results here show this - my brother and I are not completely identical.  However, we differed on only 18 base pairs out of 659,179 and 21 alleles out of the 1,318,358 tested!  So we are absolutely identical twins.  We're far closer than typical siblings can be and are closer related to one another than we are to our parents. Considering how easy it is to verify this perhaps FamilyTreeDNA should include a \"Self/Identical Twin\" category.   "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
